### PR TITLE
MNSTR-6059 Fix View History Tab Action by loading all entries in J9. …

### PR DESCRIPTION
**Pull request by ewefie**
In this PR I'm fixing issues with recently added actions: ViewHistoryTab and ViewComment.

View History Tab Action:
- in Jira 9 only the first batch of entries was loaded after clicking on 'show more' button, later the presence of the button was ignored. After fix clicking on 'show more' button is handled in the loop,
- in Jira 8 in some cases previously set duration was not enough to finish loading the entries. I increased it for all timed conditions to avoid timeout exceptions.

View Comment Action:
- improper selector was used in the function responsible for loading all comments in Jira 9,
- the previous logic for loading comments in Jira 9 was replaced with the same solution as for View History Tab - clicking 'show more' button in the loop,
- I increased the time duration in timed condition similarly as for View History Tab Action.

### CODE REVIEW

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (szymonra) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt:23]: I think Duration.ofMinutes(2) would be slightly more clear. Also if we are using this two minutes duration in most of the places I think it would be good to extract it to constant.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (szymonra) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt:90]: Is there a direct need to mark timeout and condition parameters? If not I think it will be better to stick with the convention used before - just passing the parameters. One way of another, it would be the best to stick with one convention, unless for some reason it is not sensible.

**Review by szymonra — DISMISSED** [review]

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (dagguh) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt:20]: Would a use actually wait 2 minutes?

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt
Comment (pczuj) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt:55]: We should limit users patience here, so that we won't fall into infinite loop here (e.g. button not working due to network issue). I think that after 5-10 clicks I would give up.

Comment (pczuj) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt:55]: With adding of this patience I wonder if we shouldn't fail the action in case of not loading all of the comments.

Alternatively at least maybe we should add number of comments loaded to the `observation` parameter of `meter.measure()`. This would be to make it possible to ensure that we are comparing apples to apples, e.g. if one version if Jira, doesn't cheat by loading less data.

Comment (pczuj) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt:55]: Same for `CommentTabPanel#showAllCommentsInJira9`

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (szymonra) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt]: I still think it is a good practice and while I get your point, I see this case as slightly different from what you put in the example. We switched from a specific durations to the default one, which is the same for all of the usages. However you are right that the memory/performance impact may be insignificant and if you think that it will be better to leave it the way it was, to stress out the we should approach each of the timeouts present here separately, I'm fine with that.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (ewefie) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt:20]: If they want to see all comments then they have to wait several minutes, in some cases much more than 2, but I know  that usually users give up after few seconds. If I change the value to more realistic we would important information about how our product behave under high load.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (pczuj) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt]: Yes, it's good practice to deduplicate, however I claim that in this case there was no duplication, because timeouts were different for most of the operations. Even if accidentally some of them matched it wasn't because we treated it as the same value.

We could create a common default and name it e.g. `usersPatience`. This would allow us to document rationale behind the value, because as @dagguh pointed out there was research around it. However I don't think that in our case it has enough value to balance the potential risk. Deduplicating it would suggest to ourselves and future contributors that the timeouts are meant to be manipulated all at once and not with proper granularity.

Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt]: Yeah, gotta watch out for accidental duplication. Maybe some SSH timeout is also set to `Duration.ofMinutes(2)`, but it's semantically different, they shouldn't reuse the same symbol.
And don't worry about `Duration` constructor performance, it's instant (pun intended :wink:).

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (pczuj) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt:20]: I think that we can assume that the VU who is doing the performance measures for us is rather patient for the purpose of finishing the action even if it's slow. The VU should be still have limited patience, but it's purpose is a bit different then the regular user patience, which is to not get stuck.

I'm fine with high patience of the automation as long as we have processes to monitor it (and the test does that).

**Comment by pczuj** [comment]
I'm approving with the adnotation that actions given by checkboxes in comments of this PR need to be addressed.

I believe we don't have automation to check that in GitHub.

**Review by pczuj — DISMISSED** [review]

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (dagguh) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt:20]: Reasonable justification :+1: 
Let's share it via a code comment to make it obvious to everyone.

**Comment by dagguh** [comment]
We can use the [`Require conversation resolution before merging` protection](https://github.com/atlassian/jira-actions/settings/branch_protection_rules/3773101)
image

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt
Comment (ewefie) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/CommentTabPanel.kt]: I have deduplicated duration values for both new actions, as every single step on those views seems to be very slow, but I didn't touch any other timed conditions.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt
Comment (ewefie) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt:55]: I like the idea of adding the number of loaded elements to the observation, but would it be okay for you to implement it later in a separate PR?

Comment (pczuj) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt]: @ewefie I'm specifically using "Consider..." in the action checkbox text whenever I'm ok with not doing this.

Feel free to skip this feedback if you think it's not valid for this PR.

If you think it could still be done later then you could create a [JPERF](https://ecosystem.atlassian.net/browse/JPERF) issue and address it later even by yourself.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt
Comment (pczuj) [inline src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt:58]: Nit: `for` loop with conditional `break` could be a bit more clean/safe here. Using `while` is always worrying me about deadlocks.

TIL: Apparently Kotlin doesn't have `break`, however `return` with labels can be used to achieve the same capability. This approach is compatible with other alternatives to simple `for` like `repeat()`: https://kotlinlang.org/docs/returns.html#return-to-labels

In this case it would be possible to simply `return this` in the scope of the full method - no labels needed.

File: src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt
Comment (pczuj) [reply src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/page/HistoryTabPanel.kt:43]: Maybe you even wanted to reuse the `attemptLimit` variable on the class level?

**Review by pczuj — DISMISSED** [review]

**Review by pczuj — APPROVED** [review]

**Pull request merged**
